# Oemof Workshop Week: 
## Introduction to Linear Programming and Constraints

**The concept was created by @srhbrnds in Sep 2019**

Welcome to this Notebook dealing with the basics of linear programming. We will calculate an example and interactively learn the basics of LP.

### The example: 
**LuckyBerlinCitizen has 20sqm of rooftop space. She wants to maximize her happiness (maxhp), and figured out two options to use the space. 1) If she would invest in a beautiful rooftop-terace (rt), furniture and plants, she rated her choice a 40 HappyPoints (hp) per sqm rooftop-terace. 2) Though if she would instead invest in a PV system (pv) she had a great feeling of activiely supporting the energy transformation from button-up, and rated that choice a 30 HappyPoints per sqm. Investment per sqm in the rooftop terace costs her 100 € per sqm and  investment in a PV plant costs 60 €. She only has 1500 € total investment **

#### 1) Every LP has a target funtion

From the example we derive: 

maxhp=40 x rt + 30 x pv

#### 2) Every LP has a set of constraints / inequalities

From the example we derive: 

a) Both options can never be negative  

rt &ge; 0  
pv &ge; 0

b) She has limited space on her rooftop

rt + pv &le; 20

c) She has a limited amount of money, and know that one option is more expensive than the other

100 x rt + 60 pv &le; 1500

#### 3) This reuslts in a feasible solution space 

![](./graphics/feasible_space.png)

#### 4) Solve the optimization problem

**NOTE!** In LP the solution space is infinite, as you can see. But LP obtained, that the optimal solutions are to be found in the vertices of the problem (the edges)

So in our example we only have to look at 4 points. These are:

In [1]:
import pandas as pd

edges=pd.read_csv('1_edges_pv_rt.csv')
edges

rt    pv
0  15.0   0.0
1   0.0  20.0
2   0.0   0.0
3   7.5  12.5

Now we can calculate the objective function for each pair of point (rt,pv).

In [2]:
edges['maxHP']= 40* edges['rt']+30*edges['pv']
edges

rt    pv  maxHP
0  15.0   0.0  600.0
1   0.0  20.0  600.0
2   0.0   0.0    0.0
3   7.5  12.5  675.0

Then we look for the maximum value of maxHP!

In [3]:
y = edges.loc [ edges [ 'maxHP' ] == edges [ 'maxHP' ].max() ]
y

rt    pv  maxHP
3  7.5  12.5  675.0

**Congrats you solved the LP problem :) **

LuckyBerlinCitizen should invest in 7.5 sqm roff-top terace and 12.5 sqm of pv plant to maximize her happiness!